### PR TITLE
lxc: allow creation of various resources from YAML (from Incus)

### PR DESCRIPTION
**Pull request by gabrielmougard**
cherry-picked from: https://github.com/lxc/incus/pull/786
cherry-picked from: https://github.com/lxc/incus/pull/782
cherry-picked from: https://github.com/lxc/incus/pull/778
cherry-picked from: https://github.com/lxc/incus/pull/775
cherry-picked from: https://github.com/lxc/incus/pull/767
cherry-picked from: https://github.com/lxc/incus/pull/863
cherry-picked from: https://github.com/lxc/incus/pull/864
cherry-picked from: https://github.com/lxc/incus/pull/891

### CODE REVIEW

**Comment by gabrielmougard** [comment]
I guess the cluster test suite iis broken because of the new dqlite ppa?

File: test/suites/clustering.sh [hunk 1]
```diff
@@ -3521,6 +3521,15 @@ test_clustering_groups() {
   # delete cluster group "newgroup"
   lxc cluster group delete cluster:newgroup
 
+  # Try to create a cluster group using yaml
+  lxc cluster group create cluster:node1 <<EOF
+name: newgroup-yaml
```
Comment (tomponline) [inline test/suites/clustering.sh:3526]: this shouldnt work as resource.name is used for cluster group name, not the stdin data.
Also this breaks with the convention of not allowing overriding command line args with stdin data that the other commands have.

File: lxc/storage.go [hunk 2]
```diff
@@ -103,12 +108,27 @@ func (c *cmdStorageCreate) command() *cobra.Command {
 }
 
 func (c *cmdStorageCreate) run(cmd *cobra.Command, args []string) error {
+	var stdinData api.StoragePoolPut
```
Comment (tomponline) [inline lxc/storage.go:112]: It only looks like .Config field is used from this, but what about Description?

File: test/suites/storage.sh [hunk 3]
```diff
@@ -22,6 +22,35 @@ test_storage() {
   lxc storage volume show "$storage_pool" "$storage_volume" | sed 's/^description:.*/description: bar/' | lxc storage volume edit "$storage_pool" "$storage_volume"
   lxc storage volume show "$storage_pool" "$storage_volume" | grep -q 'description: bar'
 
+  # Test creating a storage pool from yaml
+  storage_pool_yaml="lxdtest-$(basename "${LXD_DIR}")-pool-yaml"
+  if [ "${lxd_backend}" = "btrfs" ] || [ "${lxd_backend}" = "zfs" ] || [ "${lxd_backend}" = "lvm" ]; then
```
Comment (tomponline) [inline test/suites/storage.sh:27]: Can you expand the test to check for setting Description too please

**Review by tomponline — CHANGES_REQUESTED** [review]
@gabrielmougard before I go any further, please can you go through this PR (and your earlier related ones) and double check that the whole of the `Put` data being read in is actually being applied, and that the name of the resource isn't being overridden from stdin too, as it looks like at least some of these cherry-picks are only taking the Config from the Put structure and not things like Description.

File: lxc/storage.go (hunk 2 above)
Comment (gabrielmougard) [reply lxc/storage.go:112]: It seems like they forgot this case... I'll add it :) thanks

Comment (gabrielmougard) [reply lxc/storage.go:112]: fixed

File: test/suites/storage.sh (hunk 3 above)
Comment (gabrielmougard) [reply test/suites/storage.sh:27]: fixed

**Comment by gabrielmougard** [comment]
> @gabrielmougard before I go any further, please can you go through this PR (and your earlier related ones) and double check that the whole of the `Put` data being read in is actually being applied, and that the name of the resource isn't being overridden from stdin too, as it looks like at least some of these cherry-picks are only taking the Config from the Put structure and not things like Description.

It should be fixed now. I also opened this related issue: https://github.com/canonical/lxd/issues/14011

**Comment by tomponline** [comment]
@gabrielmougard did you also review your earlier PRs https://github.com/canonical/lxd/pulls?q=is%3Apr+is%3Aclosed+author%3Agabrielmougard+lxc for similar issues that I missed before?

File: lxc/storage.go [hunk 4]
```diff
@@ -124,13 +144,21 @@ func (c *cmdStorageCreate) run(cmd *cobra.Command, args []string) error {
...
+			}
+
+			pool.Config[entry[0]] = entry[1]
 		}
+	} else {
+		pool.Config = stdinData.Config
+	}
 
-		pool.Config[entry[0]] = entry[1]
+	if stdinData.Description != "" {
```
Comment (tomponline) [inline lxc/storage.go:161]: This is unnecessary, please follow the style for `cmdNetworkForwardCreate` (as an example).

The default for `Description` is empty, so we can just use it directly from stdinData.

File: lxc/snapshot.go [hunk 5]
```diff
@@ -100,6 +120,8 @@ func (c *cmdSnapshot) run(cmd *cobra.Command, args []string) error {
 
 	if c.flagNoExpiry {
 		req.ExpiresAt = &time.Time{}
+	} else if !stdinData.ExpiresAt.IsZero() {
```
Comment (tomponline) [inline lxc/snapshot.go:123]: Whats the reason we can just assign stdinData.ExpiresAt to req.ExpiresAt before line 121?

**Comment by gabrielmougard** [comment]
@tomponline is it possible to review and merge [that PR](https://github.com/canonical/lxd/pull/14016) before merging this PR?

**Comment by gabrielmougard** [comment]
> @gabrielmougard did you also review your earlier PRs https://github.com/canonical/lxd/pulls?q=is%3Apr+is%3Aclosed+author%3Agabrielmougard+lxc for similar issues that I missed before?

@tomponline After reading these past PRs, I noticed I missed adding a `description` field in the test commit of #13923 and check it has been applied... Shall I fix this in this PR? The rest of the past PRs look fine to me.

**Comment by tomponline** [comment]
> @tomponline After reading these past PRs, I noticed I missed adding a `description` field in the test commit of #13923 and check it has been applied... Shall I fix this in this PR? The rest of the past PRs look fine to me.

Yes please

File: lxc/snapshot.go (hunk 5 above)
Comment (gabrielmougard) [reply lxc/snapshot.go:123]: fixed

**Review by tomponline — APPROVED** [review]
Thanks!

**Pull request merged**